core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala

```diff
@@ -182,33 +197,55 @@ private[spark] class CoarseCookSchedulerBackend(
 
     val cmds = remoteConfFetch ++ environment ++ Seq(commandString, cleanup)
 
-    new Job.Builder()
+    val job = new Job.Builder()
       .setUUID(jobId)
       .setEnv(environment.asJava)
       .setUris(uris.asJava)
       .setCommand(cmds.mkString("; "))
-      .setMemory(sparkMesosScheduler.calculateTotalMemory(sc))
+      .setMemory(mesosBackend.calculateTotalMemory(sc))
       .setCpus(numCores)
       .setPriority(priority)
       .build()
+
+    executorsToJobs(slaveId.toString + "/" + taskId.toString) = jobId
+
+    job
   }
 
   def createRemainingJobs(): List[Job] = {
     def loop(coresRemaining: Int, jobs: List[Job]): List[Job] =
-      if (coresRemaining <= 0) jobs
-      else if (coresRemaining <= maxCoresPerJob) createJob(coresRemaining) :: jobs
-      else loop(coresRemaining - maxCoresPerJob, createJob(maxCoresPerJob) :: jobs)
-    loop(maxCores - totalCoresRequested, Nil).reverse
+      if (jobs.size == executorLimit) jobs
+      else
+        if (maxCoresPerJob <= 0) jobs
+        else if (coresRemaining <= maxCoresPerJob) createJob(coresRemaining) :: jobs
+        else loop(coresRemaining - maxCoresPerJob, createJob(maxCoresPerJob) :: jobs)
+
+    loop(maxCores - totalCoresRequested, List()).reverse
   }
 
   def requestRemainingCores() : Unit = {
     val jobs = createRemainingJobs()
-    totalCoresRequested += jobs.map(_.getCpus.toInt).sum
-    runningJobUUIDs = runningJobUUIDs ++ jobs.map(_.getUUID)
-    jobClient.submit(jobs.asJava, jobListener)
+
+    if (!jobs.isEmpty)
+      totalCoresRequested += jobs.map(_.getCpus.toInt).sum
+      runningJobUUIDs = runningJobUUIDs ++ jobs.map(_.getUUID)
+      jobClient.submit(jobs.asJava, jobListener)
+  }
+
+  override def doKillExecutors(executorIds: Seq[String]): Boolean = {
+    val uuids = executorIds.map(x => executorsToJobs(x))
+    jobClient.abort(uuids.asJavaCollection)
```

Comment:
I didn't cover this yet since I wasn't sure how to simulate a crash. Would it be OK if I refactor this a bit so that I can mock the jobClient in tests? Then I could add proper handling of errors, it's important to make sure that the executorsToJobs is updated only when we actually succeeded with aborting jobs.